extensions/ql-vscode/src/queries-panel/file-path-discovery.ts

```diff
@@ -0,0 +1,163 @@
+import { Discovery } from "../common/discovery";
+import {
+  RelativePattern,
+  Uri,
+  WorkspaceFoldersChangeEvent,
+  workspace,
+} from "vscode";
+import { MultiFileSystemWatcher } from "../common/vscode/multi-file-system-watcher";
+import { App } from "../common/app";
+import { getOnDiskWorkspaceFoldersObjects } from "../helpers";
+import { AppEventEmitter } from "../common/events";
+import { extLogger } from "../common";
+import { FilePathQueue } from "./file-path-queue";
+import { exists, lstat } from "fs-extra";
+import { filterInPlace, isEqualOrHigherPath } from "../pure/helpers-pure";
+
+interface PathData {
+  path: string;
+}
+
+/**
+ * Discovers all query files contained in the QL packs in a given workspace folder.
+ */
+export abstract class FilePathDiscovery<
+  T extends PathData,
+> extends Discovery<unknown> {
```

Comment:
Thanks for bringing this up. I should have explained this more.

I think that the `Discovery` superclass could still be useful, but it's not perfect and hence this slight misuse of it. The problem I was running into here is that the `Discovery` class provides the `discover` and `update` methods but I don't really want the `update` method. This new class doesn't operate in quite the same way of building up a big object that describes the whole state, but rather makes small updates keeps track of its own state as it goes.

The piece of behaviour that this supports is that it only calls `update` (and hence only shows any change in the UI) when there hasn't been another refresh requested. I'm not really sure why this behaviour is important, but from my point of view it seems ok to change it to always call `update` or even combine the two methods together so there's only one abstract method to override.

The behavioural effect of merging these methods would be you'd see more updates. So if a refresh is in progress and then another refresh is requested, we'd still update the UI with the first results and then update again with the second set of results. This doesn't seem like bad behaviour to me, and possibly is better because if refreshes happen a lot and discovery takes a long time then you could go a while before the `update` method is called with any new data.

The effect of the code would be one less abstract method on the `Discover` class and we remove the need for the templated type.

Overall these effects both seem good or at least neutral changes, so I'm going to go ahead and make the change. I think it's probably fine, but we could get eyes from `@dbartol` if we're not confident about it.